python/GafferUI/SpreadsheetUI/_ProxySelectionModel.py

```diff
@@ -0,0 +1,175 @@
+##########################################################################
+#
+#  Copyright (c) 2020, Cinesite VFX Ltd. All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are
+#  met:
+#
+#      * Redistributions of source code must retain the above
+#        copyright notice, this list of conditions and the following
+#        disclaimer.
+#
+#      * Redistributions in binary form must reproduce the above
+#        copyright notice, this list of conditions and the following
+#        disclaimer in the documentation and/or other materials provided with
+#        the distribution.
+#
+#      * Neither the name of John Haddon nor the names of
+#        any other contributors to this software may be used to endorse or
+#        promote products derived from this software without specific prior
+#        written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+from Qt import QtCore
+
+class _ProxySelectionModel( QtCore.QItemSelectionModel ) :
+
+	def __init__( self, viewModel, targetSelectionModel, parent = None ) :
+
+		self.__targetSelectionModel = None
+		self.__settingSelection = False
+
+		QtCore.QItemSelectionModel.__init__( self, viewModel, parent )
+
+		self.setTargetSelectionModel( targetSelectionModel )
+		self.__initMappings()
+
+		self.currentChanged.connect( self.__currentChanged )
+		self.modelChanged.connect( self.__modelChanged )
+
+	def setTargetSelectionModel( self, selectionModel ) :
```

Comment:
Just as this is a general selection proxy class, not something specific to the spreadsheet, so following our pattern of providing accessors for these things.

Reply:
Our pattern is to provide accessors for things that need accessors. For classes that are currently private and where we don't need to mutate the state, unused accessors make it harder to understand the code (where are they called, and why?) and the mutation makes for places for bugs to hide. Bear in mind that we had two such bugs already - the incorrect order to the `disconnect()` call and the `oldMode` typo that you surreptitiously fixed in https://github.com/GafferHQ/gaffer/pull/3995/commits/a2fb4a05af0ae90303cb7e8f4b03905c8f74c4c9 :) As a wise man once said, "untested code is buggy code". In this case, testing is hard because it's a UI, so I'd add "unused code paths are buggy code paths".